components/modules/ble.c

```diff
@@ -0,0 +1,1062 @@
+    message.errcode = 0;
+  }
+  if (task_block->ctxt->op == BLE_GATT_ACCESS_OP_WRITE_CHR) {
+    // Push the value
+    lua_pushlstring(L, task_block->buffer, task_block->length);
+
+    // value, struct, chrtable
+    // If we have a struct, then unpack the values
+    if (!lua_isnoneornil(L, -2)) {
+      lua_createtable(L, 0, 0);
+      int stack_size = lua_gettop(L);
+      lua_rawgeti(L, LUA_REGISTRYINDEX, struct_unpack_index);
+      lua_pushvalue(L, -4);   // dup the format
+      lua_pushvalue(L, -4);   // dup the string
+      if (lua_pcall(L, 2, LUA_MULTRET, 0)) {
+        PERROR();
+        message.errcode = BLE_ATT_ERR_UNLIKELY;
+        lua_pop(L, 2);
+        goto cleanup;
+      }
+      int vals = lua_gettop(L) - stack_size;
+
+      // Note that the last entry is actually the string offset
+      for (int i = 1; i < vals; i++) {
+        lua_pushvalue(L, -(vals - i + 1));
+        lua_rawseti(L, -(vals + 2), i);
+      }
+      lua_pop(L, vals);
+      lua_remove(L, -2);
+      // Now have table, struct, chrtable
+    }
+    // If the value is a table of a single value, then
+    // treat as value
+    if (lua_istable(L, -1) && lua_rawlen(L, -1) == 1) {
+      lua_rawgeti(L, -1, 1);
+      lua_remove(L, -2);	// and throw away the table
+    }
+    // value, struct, chr table
+    // save to `value` if present
+    lua_getfield(L, -3, "value");
+    if (!lua_isnoneornil(L, -1)) {
+      lua_pop(L, 1);
+      lua_pushvalue(L, -1);
+      lua_setfield(L, -4, "value");
+    } else {
+      lua_pop(L, 1);
+    }
+    lua_getfield(L, -3, "write");
+    if (!lua_isnoneornil(L, -1)) {
+      lua_pushvalue(L, -4);   // the characterstics table
+      lua_pushvalue(L, -3);   // the value
+      if (lua_pcall(L, 2, 0, 0)) {
+        PERROR();
+	message.errcode = BLE_ATT_ERR_UNLIKELY;
+	lua_pop(L, 2);
+	goto cleanup;
+      }
+    } else {
+      lua_pop(L, 1);  // Throw away the null write pointer
+    }
+    lua_pop(L, 1);	// THrow away the value
+    message.errcode = 0;
+  }
+
+cleanup:
+  lua_pop(L, 2);
+  message.seqno = task_block->seqno;
+
+  xQueueSend(response_queue, &message, (TickType_t) 0);
+}
+
+static int
+lble_build_gatt_svcs(lua_State *L, struct ble_gatt_svc_def **resultp, const uint16_t **handlep) {
+  // We have to first figure out how big the allocated memory is.
+  // This is the number of services (ns) + 1 * sizeof(ble_gatt_svc_def)
+  // + number of characteristics (nc) + ns * sizeof(ble_gatt_chr_def)
+  // + ns + nc * sizeof(ble_uuid_any_t)
+
+  lua_getfield(L, 1, "services");
+  if (!lua_istable(L, -1)) {
+    return luaL_error(L, "services entry must be a table");
+  }
+  int ns = lua_rawlen(L, -1);
+
+  // -1 is the services list
+  int nc = 0;
+  for (int i = 1; i <= ns; i++) {
+    MODLOG_DFLT(INFO, "Counting -- service %d (top %d)\n", i, lua_gettop(L));
+    lua_geti(L, -1, i);
+    // -1 is now the service which should be a table. It must have a uuid
+    if (lua_type(L, -1) != LUA_TTABLE) {
```

Comment:
`luaL_checktable()` would be a bit more canonical, no?

Reply:
Except that this is a nested structure of tables and I wanted to indicate *which* table was missing.